lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
+      argOperands.push_back(&argOp);
   }
+}
 
-  // Emit a destroy for each captured closure argument at each final release
-  // point.
-  for (auto *finalRelease : tracker.getFinalReleases()) {
-    builder.setInsertionPoint(finalRelease);
-    builder.setCurrentDebugScope(finalRelease->getDebugScope());
-    for (unsigned i : indices(args)) {
-      SILValue arg = args[i];
-      SILParameterInfo param = params[i];
-
-      releasePartialApplyCapturedArg(builder, loc, arg, param, callbacks);
+bool swift::collectDestroys(SingleValueInstruction *inst,
+                            SmallVectorImpl<SILInstruction *> &destroys) {
+  bool isDead = true;
+  for (Operand *use : inst->getUses()) {
+    SILInstruction *user = use->getUser();
+    if (useHasTransitiveOwnership(user)) {
+      if (!collectDestroys(cast<SingleValueInstruction>(user), destroys))
+        isDead = false;
+      destroys.push_back(user);
+    } else if (useDoesNotKeepValueAlive(user)) {
+      destroys.push_back(user);
+    } else {
+      isDead = false;
     }
   }
-
-  return true;
+  return isDead;
 }
 
-static bool
-deadMarkDependenceUser(SILInstruction *inst,
-                       SmallVectorImpl<SILInstruction *> &deleteInsts) {
-  if (!isa<MarkDependenceInst>(inst))
+/// Move the original arguments of the partial_apply into newly created
+/// temporaries to extend the lifetime of the arguments until the partial_apply
+/// is finally destroyed.
+///
+/// TODO: figure out why this is needed at all. Probably because of some
+///       weirdness of the old retain/release ARC model. Most likely this will
+///       not be needed anymore with OSSA.
+static bool keepArgsOfPartialApplyAlive(PartialApplyInst *pai,
+                                        ArrayRef<SILInstruction *> paiUsers,
+                                        SILBuilder &builder) {
+  SmallVector<Operand *, 8> argsToHandle;
+  getConsumedPartialApplyArgs(pai, argsToHandle,
+                              /*includeTrivialAddrArgs*/ false);
+  if (argsToHandle.empty())
+    return true;
+
+  // Compute the set of endpoints, which will be used to insert destroys of
+  // temporaries. This may fail if the frontier is located on a critical edge
+  // which we may not split.
+  ValueLifetimeAnalysis vla(pai, paiUsers);
+
+  ValueLifetimeAnalysis::Frontier partialApplyFrontier;
+  if (!vla.computeFrontier(partialApplyFrontier,
+                           ValueLifetimeAnalysis::DontModifyCFG)) {
     return false;
-  deleteInsts.push_back(inst);
-  for (auto *use : cast<SingleValueInstruction>(inst)->getUses()) {
-    if (!deadMarkDependenceUser(use->getUser(), deleteInsts))
-      return false;
+  }
+
+  for (Operand *argOp : argsToHandle) {
+    SILValue arg = argOp->get();
+    int argIdx = argOp->getOperandNumber() - pai->getArgumentOperandNumber();
+    SILDebugVariable dbgVar(/*Constant*/ true, argIdx);
+    builder.setInsertionPoint(pai);
```

Comment:
We are trying to move away from code like this. Can you use a new SILBuilderWithScope if possible?